lib/util.js

```diff
@@ -0,0 +1,421 @@
+ *          [propertyIsArray] true if the property is always an array, false
+ *            if not (default: false).
+ */
+api.removeValue = (subject, property, value, options) => {
+  options = options || {};
+  if(!('propertyIsArray' in options)) {
+    options.propertyIsArray = false;
+  }
+
+  // filter out value
+  const values = api.getValues(subject, property).filter(
+    e => !api.compareValues(e, value));
+
+  if(values.length === 0) {
+    api.removeProperty(subject, property);
+  } else if(values.length === 1 && !options.propertyIsArray) {
+    subject[property] = values[0];
+  } else {
+    subject[property] = values;
+  }
+};
+
+/**
+ * Compares two JSON-LD values for equality. Two JSON-LD values will be
+ * considered equal if:
+ *
+ * 1. They are both primitives of the same type and value.
+ * 2. They are both @values with the same @value, @type, @language,
+ *   and @index, OR
+ * 3. They both have @ids they are the same.
+ *
+ * @param v1 the first value.
+ * @param v2 the second value.
+ *
+ * @return true if v1 and v2 are considered equal, false if not.
+ */
+api.compareValues = (v1, v2) => {
+  // 1. equal primitives
+  if(v1 === v2) {
+    return true;
+  }
+
+  // 2. equal @values
+  if(graphTypes.isValue(v1) && graphTypes.isValue(v2) &&
+    v1['@value'] === v2['@value'] &&
+    v1['@type'] === v2['@type'] &&
+    v1['@language'] === v2['@language'] &&
+    v1['@index'] === v2['@index']) {
+    return true;
+  }
+
+  // 3. equal @ids
+  if(types.isObject(v1) && ('@id' in v1) && types.isObject(v2) && ('@id' in v2)) {
+    return v1['@id'] === v2['@id'];
+  }
+
+  return false;
+};
+
+/**
+ * Compares two strings first based on length and then lexicographically.
+ *
+ * @param a the first string.
+ * @param b the second string.
+ *
+ * @return -1 if a < b, 1 if a > b, 0 if a == b.
+ */
+api.compareShortestLeast = (a, b) => {
+  if(a.length < b.length) {
+    return -1;
+  }
+  if(b.length < a.length) {
+    return 1;
+  }
+  if(a === b) {
+    return 0;
+  }
+  return (a < b) ? -1 : 1;
+};
+
+api.promiseCallback = (fn, promise, args) => {
+  if(args.length <= fn.length) {
+    return;
+  }
+  const callback = args[args.length - 1];
+  if(typeof callback !== 'function') {
+    return;
+  }
+  return promise.then(callback.bind(null, null), callback);
+};
+
+api.callbackify = fn => {
+  return async function() {
+    const promise = fn.apply(null, arguments);
+    if(arguments.length > fn.length) {
+      const callback = Array.prototype.slice.call(arguments).pop();
+      if(typeof callback === 'function') {
+        promise.then(callback.bind(null, null), callback);
+      }
+    }
+    return promise;
+  };
+};
+
+// TODO: remove once Promise API is fixed
+/**
+ * Converts a node.js async op into a promise w/boxed resolved value(s).
+ *
+ * @param op the operation to convert.
+ *
+ * @return the promise.
+ */
+api.promisify = function(op) {
+  const args = Array.prototype.slice.call(arguments, 1);
```

Comment:
If this were to stay around, could do `api.promisify = function(op, ...args)` and omit the slice.